### DISCUSSION TITLE
Compiled Application cannot not find Specific Resources from sys.path?

### DISCUSSION THREAD

**Discussion by AMueckl** [pyinstaller/pyinstaller#7645] [Help]
Hello All,

before having learned about your discussion site, I already posted a question on stackoverflow: https://stackoverflow.com/questions/76284391/pyinstaller-compiled-application-cannot-not-find-specific-resources-from-sys-pat

There are a lot of questions towards the same problem of using sys.path for importing additional modules (like #5546, #6862), but I figured that none really fitted to the issue I got:

When trying to create a plugin feature for a larger application I came across an issue with the main.exe generated using PyInstaller on Windows  (I tried with Python 3.8 (x86) and Python 3.11 and PyInstaller 5.11.0).

The goal is to create an application that can be distributed as a package (folder) created by PyInstaller and provides a mechanism to import plugin scripts at runtime. The target machines will have a system wide Python installation, with the same version as the application was developed and compiled with. Since one cannot know in advance, which modules those plugins will need to import so that the --hidden-import cannot be used at compilation time, the application will expand the sys.path variable to include the folders of the packages of the system wide Python installation folder.

The boiled-down version of the `main.py` file would like like:

```
    import sys
    import os
    import importlib
    
    # import xmlrpc.client          # activate later on, compile, execute and compare the error messages...
    # print(dir(xmlrpc.client)[0:3])
    
    # PATH_TO_PYTHON_INSTALLATION = r'C:\Program Files (x86)\Python38'
    PATH_TO_PYTHON_INSTALLATION = r'C:\Program Files\Python311'
    PYTHONPATH = os.environ.get('PYTHONPATH', PATH_TO_PYTHON_INSTALLATION)
    LIB_FOLDERS = ['DLLs', 'Lib', '', 'Lib\\site-packages']
    for folder in LIB_FOLDERS:
        sys.path.append(os.path.join(PYTHONPATH, folder))
        pass
    
    plugin = importlib.import_module('plugin')
    print(dir(plugin.xmlrpc.server)[0:3])
```

For testing a short `plugin.py` would do:
```
    import xmlrpc.server
    print(dir(xmlrpc.server)[0:3])
```

Starting the application with the interpreter, everything works as expected.

But when the application is compiled into a `dist/main/main.exe` executable, the result will be: 

```
Traceback (most recent call last):
  File "main.py", line 13, in <module>
  File "importlib\__init__.py", line 126, in import_module
  File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1149, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "D:\pyinstaller_issue\dist\main\plugin.py", line 1, in <module>
    import xmlrpc.server
  File "C:\Program Files\Python311\Lib\xmlrpc\server.py", line 108, in <module>
    from http.server import BaseHTTPRequestHandler
ModuleNotFoundError: No module named 'http.server'
[29388] Failed to execute script 'main' due to unhandled exception!
```

The error indicates that the xmlrpc.server module was loaded from the system wide installation, but the http.server could not be imported.

It is becoming even more confusing, when one activates the import xmlrpc.client statement in the main.py, compiles again and reruns the exe-file:

```
['APPLICATION_ERROR', 'Binary', 'Boolean']
Traceback (most recent call last):
  File "main.py", line 16, in <module>
  File "importlib\__init__.py", line 126, in import_module
  File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1149, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "D:\pyinstaller_issue\dist\main\plugin.py", line 1, in <module>
    import xmlrpc.server
ModuleNotFoundError: No module named 'xmlrpc.server'
[24332] Failed to execute script 'main' due to unhandled exception!
```

I understand that in the second example the xmlrpc.client was found as an import by the pyinstaller tool and is therefore included as a frozen import into the distribution package.

What I do not understand, why now the xmlrpc.server package cannot even be found anymore (which it could before)?

**Comment by rokm** [comment]
> What I do not understand, why now the `xmlrpc.server` package cannot even be found anymore (which it could before)?

Because if a package (`xmlrpc`) is partially collected into PyInstaller's PYZ archive, it gets handled by PyInstaller's `FrozenImporter`, and the external copies that might be available in `sys.path` are not visible. 

My guess would be that `http` is also partially collected, which is why `http.server` (which is not collected) cannot be discovered from external locations.

> The target machines will have a system wide Python installation, with the same version as the application was developed and compiled with. 

Why even bother with PyInstaller, then? Just run the original code in the installed python environment...

**Reply by AMueckl** [reply]
Dear Rok,,

thank you very much for the quick response. 

> Because if a package (`xmlrpc`) is partially collected into PyInstaller's PYZ archive, it gets handled by PyInstaller's `FrozenImporter`, and the external copies that might be available in `sys.path` are not visible.
> 

I would not consider partial collection of packages making the rest of the modules unavailable being a satisfying solution. 

Why does the import need to stop at the FrozenImporter? Can one think of a possible solution making the remaining modules accessible as well?

> My guess would be that `http` is also partially collected, which is why `http.server` (which is not collected) cannot be discovered from external locations.
> 

Fair, given the logic as it currently seems to be.

> > The target machines will have a system wide Python installation, with the same version as the application was developed and compiled with.
> 
> Why even bother with PyInstaller, then? Just run the original code in the installed python environment...

Well, for example if the customer has some well-founded requirements to make it so, e.g. having control over the versions of the installed base application (as for example in software for the aerospace sector) making sure that the users are not messing around with it while still allowing for flexibility in areas that are considered not as critical.

**Reply by rokm** [reply]
> Dear Rok,,
> 
> thank you very much for the quick response.
> 
> > Because if a package (`xmlrpc`) is partially collected into PyInstaller's PYZ archive, it gets handled by PyInstaller's `FrozenImporter`, and the external copies that might be available in `sys.path` are not visible.
> 
> I would not consider partial collection of packages making the rest of the modules unavailable being a satisfying solution.
> 
> Why does the import need to stop at the FrozenImporter? Can one think of a possible solution making the remaining modules accessible as well?

Have you ever tried having a partial package available at location A and full package available at location B, and location A having precedence in search path? Even under (unfrozen) python, a regular package (non-namespace one) that's handled by built-in `SourceFileLoader` has only one search location - unless its `__path__` is externally (or internally) modified to include other locations. 

Besides, I'd say having *stdlib* split across multiple places is about as unreasonable as it can get...

On the other hand, nothing really prevents you from collecting complete stdlib into frozen application (see #6824) and avoid the issue altogether. Then you can also avoid the full python environment having to be available on the target system. For 3rd party packages, either collect them completely (`collect_submodules`), or do not collect them at all (exclude them via `--exclude-module` or equivalent argument in spec file), if you expect them to be available in external location.


> > My guess would be that `http` is also partially collected, which is why `http.server` (which is not collected) cannot be discovered from external locations.
> 
> Fair, given the logic as it currently seems to be.



> > > The target machines will have a system wide Python installation, with the same version as the application was developed and compiled with.
> > 
> > 
> > Why even bother with PyInstaller, then? Just run the original code in the installed python environment...
> 
> Well, for example if the customer has some well-founded requirements to make it so, e.g. having control over the versions of the installed base application (as for example in software for the aerospace sector) making sure that the users are not messing around with it while still allowing for flexibility in areas that are considered not as critical.

That still sounds like something that's better solved with source version control and application being installed in read-only location rather than trying to crate extensible frozen application with PyInstaller (whose main purpose is creating *self-contained* frozen applications).

**Reply by AMueckl** [reply]
> Besides, I'd say having _stdlib_ split across multiple places is about as unreasonable as it can get...
> 
I agree, but that is not intended at all. My point is that the packages are only frozen partially and the importer does not look any further if a package is there, but not completely.

> On the other hand, nothing really prevents you from collecting complete stdlib into frozen application (see #6824) and avoid the issue altogether. Then you can also avoid the full python environment having to be available on the target system. For 3rd party packages, either collect them completely (`collect_submodules`), or do not collect them at all (exclude them via `--exclude-module` or equivalent argument in spec file), if you expect them to be available in external location.

Excluding the modules really helped with the example I provided, but is not an option for the final application I am working on (which also needs to be executed in environments without a standard Python installation available).

> 
> That still sounds like something that's better solved with source version control and application being installed in read-only location rather than trying to crate extensible frozen application with PyInstaller (whose main purpose is creating _self-contained_ frozen applications).

Since to me PyInstaller still seems to be the way to go - and thanks to your suggestions - I decided to create a little package that can compare the modules that would be frozen against the ones available in the standard Python installation and iteratively extends the hidden-import list times until every submodule of all packages are included if there was a partial import already.

So I decided to create a helping package ([https://pypi.org/project/wrapenv/](https://pypi.org/project/wrapenv/)) to be able to tweak the Analysis() step of the build process as shown in the example file: [https://github.com/AMueckl/WrapEnv/blob/main/examples/run_pyinstaller.py](https://github.com/AMueckl/WrapEnv/blob/main/examples/run_pyinstaller.py)
